### PR TITLE
[WIP] Xenial   android9   addremovecontext

### PR DESCRIPTION
**Pull request by Pekkari**
This is a WIP PR to allow testing over ubports-qa of a solution to issues:

https://github.com/HelloVolla/ubuntu-touch-beta-tests/issues/94
https://github.com/HelloVolla/ubuntu-touch-beta-tests/issues/105

I'll wave up when the test is over.

### CODE REVIEW

File: ofono/src/gprs.c [hunk 1]
```diff
@@ -854,9 +855,10 @@ static void pri_reset_context_settings(struct pri_context *ctx)
 
 	pri_context_signal_settings(ctx, signal_ipv4, signal_ipv6);
 
-	if (ctx->type == OFONO_GPRS_CONTEXT_TYPE_MMS) {
```
Comment (Flohack74) [inline ofono/src/gprs.c:857]: Where is the matching closing } bracket? Why is this removed?

Comment (Pekkari) [reply ofono/src/gprs.c:857]: This is coming from commit e52746aed, if you read it, the opening brace is added in a new
if clause a few lines under.

**Comment by NotKit** [comment]
@Pekkari have you checked if it's possible to remove extra APNs from settings, keeping only Internet+MMS one, so the extra ones do not reappear after reboot? Reappearing issue seems to remain according to HelloVolla/ubuntu-touch-beta-tests#94.

**Comment by Pekkari** [comment]
@NotKit, I haven't, but I can tell that it may be possible to remove only the mms one, since
the combined shows as the internet one, the interface is not smart to detect when it is internet
context or combined, for what we can work in another PR in system settings. I'll do the test though.

File: ofono/src/gprs.c (hunk 1 above)
Comment (Flohack74) [reply ofono/src/gprs.c:857]: Ah Im stupid yes

**Comment by Pekkari** [comment]
@NotKit for me the extra APNs keep reappearing as long as there is any entry in
/usr/share/mobile-broadband-provider-info/serviceproviders.xml that can take up
the mms APN(even when it is not my operator). Removing the extra APNs let me
use the internet, but not the mms, so either my tests weren't ok, or the code may
require further improvements. I still give room to believe DNA Finland doesn't work in
combined mode though.

**Comment by Pekkari** [comment]
I restored the fix here so that people can keep testing and moved the new code to the branch xenial_-_android9_-_addremovecontext_backup, so that developers can check the code, and share thoughts. Please feel free to do your tests, and if it breaks, this should behave as well as formerly.

**Comment by fredldotme** [comment]
I will let a few US users with particular interest in this topic test this if you don't mind.
Do you, from your PoV, see this PR as complete? @Pekkari

**Comment by stekalb** [comment]
I can test it on my Volla at the next weekend

**Comment by Pekkari** [comment]
@fredldotme testing from other user reports it is in bad shape, I'd like
to hear more experiences since I have succeeded in restarting ofono
without the extra context. Otherwise I would consider it finished yes.

If things goes wrong on your end, I'll remove the last patch to let you
recover the original incomplete fix.

**Comment by rik-shaw** [comment]
Reference: install this patch with this command: `ubports-qa install PR_ofono_28`

I am one of the US users that have been testing, and in my experience if I set an APN to "Internet+MMS" in the GUI, it does cause MMS and Internet to work correctly with a single APN (previously we needed 2 different APNs: one for internet and a different one for MMS, and each would need different APN values).

However, after setting "Internet+MMS" in the GUI, if I then look at `/var/lib/ofono/####/gprs` it seems that the contexts have been deleted, and in fact if I reboot then the `gprs` file is getting set back to default entries. So the "stopgap" solution for me currently is to edit `/usr/share/mobile-broadband-provider-info/serviceproviders.xml` with the values for my carrier (US Mobile, a T-Mobile mvno in the USA). Then when gprs is wiped out and reset, it brings in the correct settings for my carrier and even after reboot Internet + MMS works. I comment out all other items for my carrier information.

NOTE that in `serviceproviders.xml` I use `<usage type="internet">` along with the correct `<mmsc>` entry. I do NOT see other places where `usage type="internet"` and a `mmsc` entry are used together, but in practice that is the key to allowing the single APN to be seen as "Internet+MMS"

NOTE that sometimes after reboot, the `gprs` file still doesn't have the context entries, and MMS / Cellular Data don't appear to work. If I go in the GUI to "APN" and untick the "US Mobile" entry, it then "auto-re-enables" after being unticked, and then MMS and Internet work again. gprs file still doesn't have the context entries in it, but the system knows about it and things work :-)

For reference, here is the `serviceproviders.xml` entry for US Mobile (if you search for `epc` you will find the US T-Mobile entries in the file):
```
    <apn value="pwg">
       <usage type="internet"/>
       <name>US Mobile</name>
       <mmsc>http://pwg.mmsmvno.com/mms/wapenc</mmsc>
    </apn>
```

PS: for me it was essential to do the above settings since when using 2 APNs (one for Internet, a second for MMS) I didn't have 3G cellular data (only 4G was working). The combined APN has working cellular data in 3G and 4G mode as well as functioning MMS.

**Comment by jackmcdougal** [comment]
My experience is as follows:  From the UI, I can not create a combined APN ... after entering the values, I click the top right check box, it goes grey, and nothing else happens.  Back out to list of APNS, and it shows that a new APN was created, but type=Internet and none of the mms settings are available or saved.  If I change the serviceproviders.xml file as per comment above, delete all apns, then reboot (or restart ofono), I end up with a good working combiined (Internet & MMS) apn that holds through reboots.  Depending on how I modify my carrier's serviceprovider.xml entry, extra apn's may or may not be added (on reboot or ofono restart). For example, if I leave only the internet apn with the mmsc info (the one that creates the merged single entry) in the xml file, It will also create an empty mms entry in the UI that is unchecked.  If I delete the whole section for my carrier, I end up with an empty internet and an empty mms.   On a probably unrelated note but mentioning in case, my inbound mms is failing with the oops msg:  Nuntium log shows a working download URL (works via wget), but finishes with a DBUS message about an invalid URL: 
00000030  33 33 38 34 62 34 35 61  66 36 65 33 31 31 63 63  |3384b45af6e311cc|
00000040  66 37 65 36 30 62 37 2d  31 00 8d 92 89 1a 80 18  |f7e60b7-1.......|
00000050  ea 2b 31 35 31 34 32 30  39 36 32 35 39 2f 54 59  |.+15142096259/TY|
00000060  50 45 3d 50 4c 4d 4e 00  96 00 8a 80 88 05 81 03  |PE=PLMN.........|
00000070  09 3a 80 86 81 83 68 74  74 70 3a 2f 2f 6d 6d 73  |.:....http://mms|
00000080  63 32 2e 67 65 74 2e 62  65 6c 6c 2e 63 61 2f 63  |c2.get.bell.ca/c|
00000090  6c 6f 75 64 6d 6d 73 63  3f 6c 3d 31 33 33 38 34  |loudmmsc?l=13384|
000000a0  62 34 35 61 66 36 65 33  31 31 63 63 66 37 65 36  |b45af6e311ccf7e6|
000000b0  30 62 37 2d 31 00 8e 03  04 35 31                 |0b7-1....51|
2021/12/09 21:02:06 Download issues: org.freedesktop.DBus.Error.InvalidArgs: Invalid URL: ''

env:
dev build 
pixel 3a
ubports-qa install PR_ofono_28
carrier bell canada

gprs 
[context4]
Name=Internet
AccessPointName=pda.bell.ca
Username=
Password=
AuthenticationMethod=none
Type=internet
Protocol=ip
Preferred=true
MessageProxy=web.wireless.bell.ca
MessageCenter=http://mms.bell.ca/mss/wapenc

I have also tried with the port on the proxy: 
MessageProxy=web.wireless.bell.ca:80

(When the port is in there, there is also an odd UI bug: when the merged apn is opened, it first shows the port at the end of the proxy field, but then it quickly jumps into the proxy port field).

hope this helps and thanks for your work!!

**Comment by Pekkari** [comment]
Thanks for the testing and comments! @jackmcdougal that UI issue is known to me, I usually stay
away from it by removing the internet context and modifying the mms to point to the internet access
point name. Of course, remind to set the mms context to internet + mms. I hope it helps!

**Comment by fredldotme** [comment]
I would highly appreciate if setting Internet + MMS APNs was properly fixed before merging it.
Since so many users are affected by the bug this intends to fix, the UI part is crucial in making this work.
It might just be that the code responsible for saving the gprs file is not taking it into account yet.

**Comment by rik-shaw** [comment]
> I would highly appreciate if setting Internet + MMS APNs was properly fixed before merging it. Since so many users are affected by the bug this intends to fix, the UI part is crucial in making this work. It might just be that the code responsible for saving the gprs file is not taking it into account yet.

I suspect (??) that the problem with creating the new APN and its settings are not saved is the same as issue #24 meaning that it isn't a regression caused by this code?

However, the "deleting of all contexts" in the gprs file is more of a concern, and would be unwise to release with this issue. Any thoughts on where to look into this problem? Maybe by setting "Type=internet" plus having a "MessageCenter=" value entered it is causing an unexpected crash somewhere (even though that combination in practice seems to make it work correctly)?

**Comment by rik-shaw** [comment]
As of devel 2021-12-17 I can confirm that without this `ofono 28` patch that I can not download MMS in 3G mode using US Mobile (T-Mobile mvno) in the USA. This could be because with  needing 2 different APNs (one for MMS, another for Internet) the APN I am using for "Internet" is the upstream T-Mobile settings and they may have (??) turned off 3G?? The combined "Internet + MMS" US Mobile APN could work because of being able to use that same network.

**Comment by gbdomubpkm** [comment]
Well. I have a new (maybe) info : I don't know if it's a good news or not, but it seems that the change of the user interface is not directly related to the PR -28 (observation of non-developer user) :

On OTA-20, I wanted to install the gpu browser on my Volla (gpu browser works well on it), as I did in previous OTAs. The procedure went well, except that after reboot, I found myself with the new presentation of the lock screen (4 circles for the pin code) and again to the access â the gallery blocked (blue screen with the mountains frozen) as at the time of my installation of the last version of the PR-28 that does not work.

Fortunately, I was able to switch to RC 42 through the system settings (and avoid a new flashing with wipe), and access to the gallery is possible again. The new unlock screen is still there, I don't know if this is normal.

**Comment by Pekkari** [comment]
> Well. I have a new (maybe) info : I don't know if it's a good news or not, but it seems that the change of the user interface is not directly related to the PR -28 (observation of non-developer user) :
> 
> On OTA-20, I wanted to install the gpu browser on my Volla (gpu browser works well on it), as I did in previous OTAs. The procedure went well, except that after reboot, I found myself with the new presentation of the lock screen (4 circles for the pin code) and again to the access â the gallery blocked (blue screen with the mountains frozen) as at the time of my installation of the last version of the PR-28 that does not work.

I suspect your problem is the service provider database, look how it looks like for your operator:
```
        <provider>
                <name>Bouygues Telecom</name>
                <gsm>
                        <network-id mcc="208" mnc="20"/>
                        <network-id mcc="208" mnc="21"/>
                        <network-id mcc="208" mnc="88"/>

                        <voicemail>660</voicemail>
                        <balance-check>
                                <dtmf>630</dtmf>
                        </balance-check>

                        <apn value="a2bouygtel.com">
                                <plan type="postpaid"/>
                                <usage type="internet"/>
                                <name>Contrat Pro Data Illimité</name>
                        </apn>
                        <apn value="b2bouygtel.com">
                                <plan type="postpaid"/>
                                <usage type="internet"/>
                                <name>B2Bouygtel</name>
                        </apn>
                        <apn value="ebouygtel.com">
                                <plan type="postpaid"/>
                                <usage type="internet"/>
                                <name>Forfait Data</name>
                        </apn>
                        <apn value="mmsbouygtel.com">
                                <plan type="postpaid"/>
                                <usage type="internet"/>
                                <name>Forfait Data</name>
                        </apn>
                        <apn value="pcebouygtel.com">
                                <plan type="postpaid"/>
                                <usage type="internet"/>
                                <name>Contrat Pro Data</name>
                        </apn>
                        <apn value="mmsbouygtel.com">
                                <usage type="mms"/>
                                <name>Bouygues MMS</name>
                                <mmsc>http://mms.bouyguestelecom.fr/mms/wapenc</mmsc>
                                <mmsproxy>62.201.129.226:8080</mmsproxy>
                        </apn>
                </gsm>
        </provider>
```
Not only comes as a separate apn, it also provides multiple choices for internet and mms,
this can be modified under /usr/share/mobile-broadband-provider-info/service-provider.xml.
If we find together a good fit for you, I can upstream a patch to sailfish to fix it.

**Comment by rik-shaw** [comment]
@Pekkari I have settings for US Mobile (a T-Mobile mvno in the USA) that I would love to get into `serviceproviders.xml`. Can you help me know how to get ahold of the `mcc` and `mnc` values so I can make sure to format the US Mobile entry correctly?

**Comment by Pekkari** [comment]
if you set your device in developer mode, and access it through adb shell, you can execute
the included ofono script list-operators located at /usr/share/ofono/scripts. It's output would
look like:
```
# ./list-operators 
[ /ril_1 ]
[ /ril_0 ]
    [ /ril_0/operator/24412 ]
        MobileNetworkCode = 12
        MobileCountryCode = 244
        Status = current
        Name = DNA
        Technologies = lte
```

**Comment by rik-shaw** [comment]
@Pekkari here is the info for US Mobile in the USA (T-Mobile mvno):

```
        <provider>
                <name>US Mobile</name>
                <gsm>
                        <network-id mcc="310" mnc="260"/>
                        <voicemail>123</voicemail>
                        <apn value="pwg">
                                <usage type="internet"/>
                                <name>US Mobile</name>
                                <mmsc>http://pwg.mmsmvno.com/mms/wapenc</mmsc>
                        </apn>
                </gsm>
        </provider>
```

As part of a patch, the 310 / 260 entry would need to be removed from the T-Mobile provider info. I could try making the PR but you mentioned it is best done at sailfish? A bit confused where I would go to make a PR. Thanks again for the tips!

**Comment by Pekkari** [comment]
I'm afraid so, ubports snapshots a database coming from sailfish, as well as it snapshots their ofono, [here](https://github.com/sailfishos/mobile-broadband-provider-info/blob/master/mobile-broadband-provider-info/serviceproviders.xml) is the source.

**Comment by Flohack74** [comment]
@rik-shaw please check first https://gitlab.gnome.org/GNOME/mobile-broadband-provider-info - that should be the ultimate upstream for all projects, if they have correct data or not.

**Comment by rik-shaw** [comment]
> @rik-shaw please check first https://gitlab.gnome.org/GNOME/mobile-broadband-provider-info - that should be the ultimate upstream for all projects, if they have correct data or not.

@Flohack74 thank you, I have submitted a merge request for US Mobile here: https://gitlab.gnome.org/GNOME/mobile-broadband-provider-info/-/merge_requests/54

**Comment by Flohack74** [comment]
@rik-shaw Thanks, hope they accept it. I think we should pull this db from GNOME rather than any of the downstreams in between as we will otherwise always have heavily outdated data. The disadvantage if you repackage already repackaged things ;) - copuld you open a ticket for this here in ofono, I try to take care of this.

**Comment by rik-shaw** [comment]
> @rik-shaw Thanks, hope they accept it. I think we should pull this db from GNOME rather than any of the downstreams in between as we will otherwise always have heavily outdated data. The disadvantage if you repackage already repackaged things ;) - copuld you open a ticket for this here in ofono, I try to take care of this.

Done: Issue #33 . I think you are right about keeping updated, there are 18 merge requests pending at the GNOME project :-)

**Comment by rik-shaw** [comment]
@Pekkari sorry for all the "noise" in this PR with talk of how to add to Mobile Broadband Provider Info, etc. But back to this PR, again I am running it and it does allow the combined APN which is great! But the question of it regularly resetting the `/var/lib/ofono/###/gprs` file, is that something you see also?

Maybe after OTA-21 is released then it is better to re-test all of this? It would be GREAT to see this PR merged in, hopefully OTA-22? :-) Thanks again for the great work.

**Comment by gbdomubpkm** [comment]
> regularly resetting the /var/lib/ofono/###/gprs file.

With less technical words, when i had APN's configuration lost/changed with Bouygues Sim after reboot with an attempt with this PR-28, could it be an effect of that ?

**Comment by Pekkari** [comment]
> @Pekkari sorry for all the "noise" in this PR with talk of how to add to Mobile Broadband Provider Info, etc. But back to this PR, again I am running it and it does allow the combined APN which is great! But the question of it regularly resetting the `/var/lib/ofono/###/gprs` file, is that something you see also?

I personally haven't noticed that, and I need to reinstall the PR to check. What is the file about? I would
assume anyways anything under /var/lib/ can be modified by the code at will so honestly I see no real
problem with that overrriding it, unless the override is done wrong and it produces some side effect on
your device, in which case, we should chase the code that wrongly override it.

**Comment by rik-shaw** [comment]
> I personally haven't noticed that, and I need to reinstall the PR to check. What is the file about
That is the text file that holds the APN settings, and what I assume is read by ofono on boot or cellular data enable and written to on APN settings change, so it being wiped out means after reboot any saved custom APN settings were removed.

HOWEVER, I am re-testing and do not see this file getting erased on boot, maybe something else in the ofono / etc MMS updates was causing this problem before.

So, at least from my side I think this one is working well and it would be great to see if it could make it for OTA-22. @fredldotme is there any blocker you know of or could this be considered for addition to OTA-22?

**Comment by rik-shaw** [comment]
I did some clean testing (reverted to stable then to devel again and reset all APNs), and  unfortunately I can confirm my prior problematic observations: Something is resetting `/var/lib/ofono/###/gprs` to default values from `serviceproviders.xml` so any custom APN settings are getting reset / removed often. Not sure yet what the trigger is: if it is on every boot it is reset, if it is changing some APN values that causes the reset, or other. Will need more testers?

**Comment by rik-shaw** [comment]
@Flohack74 could this be considered for OTA-22? There is something strange with `gprs` maybe getting reset but other testers may prove it is not a problem and is "only me". @Pekkari do you think this PR is complete and ready to be considered for merging from your view?

**Comment by Pekkari** [comment]
> @Flohack74 could this be considered for OTA-22? There is something strange with `gprs` maybe getting reset but other testers may prove it is not a problem and is "only me". @Pekkari do you think this PR is complete and ready to be considered for merging from your view?

I can reproduce a problem with the mms context reappearing any 2nd reboot, which is a blocker to land this,
however, finding a solution for that is a crusade :\

**Comment by rik-shaw** [comment]
> I can reproduce a problem with the mms context reappearing any 2nd reboot, which is a blocker to land this, however, finding a solution for that is a crusade :\

OK well at least we see the same thing :-) I think that somewhere in the code there is logic that "if it doesn't seem that the APN settings in `gprs` are valid, then reset. Maybe by having the combined APN (with the only way to do that being have MMSC filled out and type of internet) it wrongly thinks it is invalid? I did note in `serviceproviders.xml` I don't think there are ANY other settings that have MMSC filled out plus type of internet.

So @Flohack74 I think this means it is not a candidate for OTA-22 :-(

**Comment by Flohack74** [comment]
No, probably not a candidate, needs more testing also with other devices...

**Comment by JamiKettunen** [comment]
For users needing this on 20.04 https://gitlab.com/ubports/development/core/packaging/ofono-sailfish/-/merge_requests/11 exists which you can get on the development update channel using:
```
sudo ubports-qa install focal_-_PR_ofono-sailfish_11
```
It'll need further investigation though so it might not end up getting merged in the end, but at least it exists for installation so those who desperately need it have it available in some form :)

**Comment by rik-shaw** [comment]
@JamiKettunen thanks for summarizing and for clarifying. I'll see if there are some USA users that would like to test it out. Again one difference for us (USA) is that we don't need to specify a port, which is what Lionel Duboeuf needs in his location (France?). Regardless, by not needing 2 APNs it may simplify greatly the usage for USA users (we still can only receive calls in 2G mode however, but another issue waiting on VoLTE support :-)

**Comment by gbdomubpkm** [comment]
The PR-28 hasn't worked for me at all since the March 16 RC update (I've been in RC for over two years without changing channels and it was ok before march 16 update). I uninstalled it and the effect is the same: to be able to use MMS I have to install 2 different APNs like two years ago.

**Comment by peat-psuwit** [comment]
Hello,

Thank you for contributing to UBports. As part of project renaming and the effort to port Ubuntu Touch stack to Ubuntu 20.04, we're incrementally migrating repositories to GitLab.

This PR has been further worked upon on GitLab [1] and is now merged. As such, I'll close this PR.

[1] https://gitlab.com/ubports/development/core/packaging/ofono-sailfish/-/merge_requests/13
